### PR TITLE
Dim reduction modularity

### PR DESCRIPTION
**Pull request by pabloT97**
Add dimensionality reduction module and tests

### CODE REVIEW

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+        self.dataset = dataset
+        if 'sentence_embedding' not in self.dataset:
+            raise ValueError('sentence vector not present in the dataset')
+        self.sentence_vectors = dataset['sentence_embedding'].to_list()
+        self.dim_out = dim_out
+
+    @abstractmethod
+    def fit(self) -> None:
+        """
+        Fits the dataset to output vectors with the desired dimension
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py:46]: Please change the name!
Functions called 'fit' have a very specific purpose and its not that one. It's very confusing for someone to coma accross a function called fit that isn't about optimisation.

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    def save_dataset(
+            self, filename :str,
+            dataset_path : str) -> None:
+        """
+        Saves the reduced dataset in the specified directory
+
+        Parameters
+        ----------
+        filename : str
+            Name of the saved file
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py:58]: Of the saved file or of the file to save to?

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+                Arguments passed to the sklearn.decomposition.PCA object
+                They can be found in 
+                https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html
+        """
+        super().__init__(
+            dataset = dataset, dim_out = dim_out
+        )
+        self.pca_sk = skd.PCA(n_components=self.dim_out, **kwargs)
+    
+    def fit(self):
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py:97]: same as above

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+                https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FastICA.html
+        """
+        super().__init__(
+            dataset = dataset, dim_out = dim_out
+        )
+        self.ica_sk = skd.FastICA(
+            n_components = self.dim_out, **kwargs
+        )
+
+    def fit(self):
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py:137]: same

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        
+class TSVD(DimReduction):
+    """
+    Class for truncated SVD dimensionality reduction
+    """
+    def __init__(
+        self, dataset : pd.DataFrame, dim_out : int, **kwargs
+    )-> None:
+        """
+        Initialises the tSVD dimensionality reduction class
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py:156]: why is the t small in tSVD?

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+                https://umap-learn.readthedocs.io/en/latest/parameters.html
+        """
+        super().__init__(
+            dataset = dataset, dim_out = dim_out
+        )
+        self.umap_sk = umap.UMAP(
+            n_components = self.dim_out, **kwargs
+        )
+
+    def fit(self):
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/dim_reduction.py:219]: same

File: tests/test_dim_reduction.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+import numpy as np 
+import os 
+import sys 
+current_path = os.path.dirname(os.path.abspath(__file__))
+sys.path.append(current_path + "/../neasqc_wp61/models/quantum/")
+from dim_reduction import *
+
+class TestDimReduction(unittest.TestCase):
+    
+    def setUp(self):
```
Comment (ryk-wolf) [inline tests/test_dim_reduction.py:12]: the dimension (here 768) should probably be an argument and not hard-coded. We want to re-use this for other embeddings which might be of different sizes.
Always make things generic when you can.

File: tests/test_dim_reduction.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        Generate random dataset for our tests
+        """
+        n_rows = 10
```
Comment (ryk-wolf) [inline tests/test_dim_reduction.py:16]: Not sure what n_rows means here... is it the number of fake sentences we want? If so, pls find a better variable name. Also then it should be a parameter and not hard-coded.

File: tests/test_dim_reduction.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        n_rows = 10 
+        data = {
+            'label' : np.random.randint(2, size=n_rows),
+            'sentence_embedding' : [
+                np.random.rand(768) for _ in range(n_rows)]
+        }
+        self.df = pd.DataFrame(data)
+
+
+    def testPCA(self):
```
Comment (ryk-wolf) [inline tests/test_dim_reduction.py:25]: This isn't how one names tests.
Tests should have explicit names, ie you should tell what you're testing for. Something like:

PCAOutputsDimensionNEmbeddingsWhenAskedForIt

Like, I should know exactly WHAT the test is testing just by reading the name, without the code or any comments.

File: tests/test_dim_reduction.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        Test of PCA dimensionality reduction. 
+        """
+        dim_out = 4
```
Comment (ryk-wolf) [inline tests/test_dim_reduction.py:29]: The output dimension should NOT be hard-coded, it should be a parameter that you pass.
Also, make sure to test with a number of different parameters, there are ways to do that. You give a list of dimensions as input to your test, and it will be tested for each of those inputs.
We don't want out function to particularly always output an embedding of dim 4, we want it to output a dim N when we ask for a dim N reduction. Whatever that N may be.
 This is what your test should reflect.

File: tests/test_dim_reduction.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        PCA_object = PCA(self.df, dim_out, svd_solver = 'full', tol = 9.0)
+        PCA_object.fit()
+        for value in PCA_object.dataset['reduced_sentence_embedding']:
+            self.assertEqual(len(value), dim_out)
+    
+    def testICA(self):
```
Comment (ryk-wolf) [inline tests/test_dim_reduction.py:35]: same remarks as for PCA test

File: tests/test_dim_reduction.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        Test of ICA dimensionality reduction. 
+        """
+        dim_out = 2
+        ICA_object = ICA(self.df, dim_out, fun = 'cube')
+        ICA_object.fit()
+        for value in ICA_object.dataset['reduced_sentence_embedding']:
+            self.assertEqual(len(value), dim_out)
+
+    def testTSVD(self):
```
Comment (ryk-wolf) [inline tests/test_dim_reduction.py:45]: same remarks as for PCA test

File: tests/test_dim_reduction.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        Test of TSVD dimensionality reduction. 
+        """
+        dim_out = 2
+        TSVD_object = TSVD(self.df, dim_out, algorithm = 'arpack')
+        TSVD_object.fit()
+        for value in TSVD_object.dataset['reduced_sentence_embedding']:
+            self.assertEqual(len(value), dim_out)
+
+    def testUMAP(self):
```
Comment (ryk-wolf) [inline tests/test_dim_reduction.py:55]: same remarks as for PCA test

File: tests/test_dim_reduction.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        Test of UMAP dimensionality reduction. 
+        """
+        dim_out = 4
+        UMAP_object = UMAP(self.df, dim_out, n_neighbors=2)
+        UMAP_object.fit()
+        for value in UMAP_object.dataset['reduced_sentence_embedding']:
+            self.assertEqual(len(value), dim_out)  
+
+    def testTSNE(self):
```
Comment (ryk-wolf) [inline tests/test_dim_reduction.py:65]: same remarks as for PCA test

File: tests/test_dim_reduction.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        Test of TSNE dimensionality reduction. 
+        """
+        dim_out = 3
+        TSNE_object = TSNE(self.df, dim_out, perplexity=5.0)
+        TSNE_object.fit()
+        for value in TSNE_object.dataset['reduced_sentence_embedding']:
+            self.assertEqual(len(value), dim_out)  
+
+    def test_save_dataset(self):
```
Comment (ryk-wolf) [inline tests/test_dim_reduction.py:75]: Same as above for the name, I should know WHAT behaviour you're testing for

File: tests/test_dim_reduction.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+        """
+        Test save dataset function
+        """
+        dim_out = 8
+        UMAP_object = UMAP(self.df, dim_out, n_neighbors=2)
```
Comment (ryk-wolf) [inline tests/test_dim_reduction.py:80]: I'm confused, why are we testing only with a UMAP object? 🤔 don't we want to make sure this works with any dimensionality reduction technique?

**Review by ryk-wolf — CHANGES_REQUESTED** [review]
Good code overall. Quite a few changes to be made still.

File: neasqc_wp61/models/quantum/dim_reduction.py (hunk 5 above)
Comment (pabloT97) [reply neasqc_wp61/models/quantum/dim_reduction.py:156]: mistake

File: tests/test_dim_reduction.py (hunk 16 above)
Comment (pabloT97) [reply tests/test_dim_reduction.py:80]: This is because save_dataset method on parent class, so it is the same in all child objects. That is why it's only tested with UMAP

**Review by ryk-wolf — CHANGES_REQUESTED** [review]
Much better overall! Keep it up :)

File: neasqc_wp61/models/quantum/dim_reduction.py [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+    
+    def reduce_dimension(self)-> None:
+        """
+        Fit the vectorised sentences to obtain the reduced dimension
+        sentence vectors.
+        """
+        sentence_vectors_reduced = self.tsne_sk.fit_transform(
+            np.array(self.sentence_vectors)
+        )
+        self.reduced_dataset = self.dataset.copy()
```
[the remark this answers is not in the document]
Comment (pabloT97) [reply neasqc_wp61/models/quantum/dim_reduction.py:276]: Line 276 has been left there from previous code and I forgot to remove it, sorry about that.. I will remove it rn, no need to copy the datasets.

**Review by ryk-wolf — APPROVED** [review]
Great job!
